examples/applications/plot_cyclical_feature_engineering.py

```diff
@@ -0,0 +1,741 @@
+
+"""
+==========================
+Cyclic feature engineering
+==========================
+
+This notebook introduces different strategies to leverage time-related features
+for a bike sharing demand regression task that is highly dependent on business
+cycles (days, weeks, months) and yearly season cycles.
+
+In the process, we introduce how to perform periodic feature engineering using
+the :class:`sklearn.preprocessing.SplineTransformer` class and its
+`extrapolation="periodic"` option.
+
+"""
+# %%
+# Data exploration on the Bike Sharing Demand dataset
+# ---------------------------------------------------
+#
+# We start by loading the data from the OpenML repository.
+from sklearn.datasets import fetch_openml
+
+bike_sharing = fetch_openml("Bike_Sharing_Demand", version=2, as_frame=True)
+df = bike_sharing.frame
+
+# %%
+# To get a quick understanding of the periodic patterns of the data, let us
+# have a look at the average demand per hour during a week.
+#
+# Note that the week starts on a Sunday, during the weekend. We can clearly
+# distinguish the commute patterns in the morning and evenings of the work days
+# and the leisure use of the bikes on the weekends with a more spread peak
+# demand around the middle of the days:
+import matplotlib.pyplot as plt
+
+
+average_week_demand = df.groupby(["weekday", "hour"]).mean()["count"]
+average_week_demand.plot(figsize=(12, 4))
+_ = plt.title("Average number of bike rentals during the week")
+
+
+# %%
+#
+# The target of the prediction problem is the absolute count of bike rentals on
+# a hourly basis:
+df["count"].max()
+
+# %% [markdown]
+#
+# Let us rescale the target variable (number of hourly bike rentals) to predict
+# a relative demand so that the mean absolute error is more easily interpreted
+# as a fraction of the maximum demand.
+#
+# .. note::
+#
+#     The fit method of the models used in this notebook all minimize the
+#     mean squared error to estimate the conditional mean instead of the mean
+#     absolute error that would fit an estimator of the conditional median.
+#
+#     When reporting performance measure on the test set in the discussion, we
+#     instead choose to focus on the mean absolute error that is more
+#     intuitive than the (root) mean squared error. Note however that the best
+#     models for one metric are also the best for the other in this study.
+y = df["count"] / 1000
+_ = y.hist(bins=30)
```

Comment:
It could be nice to have some `xlabel` on the histogram.

Reply:
done.